packages/flutter/lib/src/painting/image_resolution.dart

```diff
@@ -34,18 +38,52 @@ const String _kAssetManifestFileName = 'AssetManifest.json';
 ///
 /// For example, suppose an application wants to use an icon named
 /// "heart.png". This icon has representations at 1.0 (the main icon), as well
-/// as 1.5 and 2.0 pixel ratios (variants). The asset bundle should then contain
-/// the following assets:
+/// as 2.0 and 4.0 pixel ratios (variants). The asset bundle should then
+/// contain the following assets:
 ///
 /// ```
 /// heart.png
-/// 1.5x/heart.png
 /// 2.0x/heart.png
+/// 4.0x/heart.png
 /// ```
 ///
 /// On a device with a 1.0 device pixel ratio, the image chosen would be
-/// heart.png; on a device with a 1.3 device pixel ratio, the image chosen
-/// would be 1.5x/heart.png.
+/// heart.png; on a device with a 2.0 device pixel ratio, the image chosen
+/// would be 2.0x/heart.png; on a device with a 4.0 device pixel ratio, the
+/// image chosen would be 4.0x/heart.png.
+///
+/// On a device with a device pixel ratio that does not exactly match an
+/// available asset the "best match" is chosen. Which asset is the best depends
+/// on the screen. Low-resolution screens (those with device pixel ratio
+/// strictly less than 2.0) use a different matching algorithm from the
+/// high-resolution screen. Because in low-resolution screens the physical
+/// pixels are visible to the user upscaling artifacts (e.g. blurred edges) are
+/// more pronounced. Therefore, a higher resolution asset is chosen, if
+/// available. For higher-resolution screens, where individual physical pixels
+/// are not visible to the user, the asset variant with the pixel ratio that's
+/// the closest to the screen's device pixel ratio is chosen.
+///
+/// For example, for a screen with device pixel ratio 1.25 the image chosen
+/// would be 2.0x/heart.png, even though heart.png (i.e. 1.0) is closer. This
+/// is because the screen is considered low-resolution. For a screen with
+/// device pixel ratio of 2.25 the image chosen would also be 2.0x/heart.png.
+/// This is because the screen is considered to be a high-resolution screen,
+/// and therefore upscaling a 2.0x image to 2.25 won't result in visible
+/// upscaling artifacts. However, for a screen with device-pixel ratio 3.25 the
+/// image chosen would be 4.0x/heart.png because it's closer to 4.0 than it is
+/// to 2.0.
```

Comment:
Doesn't this depend on how much upscaling is happening? Like upscaling from 1.0 to 1.25 is a 25% increase, but 2.0 to 2.25 is only a 12.5% increase.

I'm just throwing an idea out there in case it makes sense to you: what if, instead of having a 2.0 DPR limit, we check the upscaling factor and decide based on that. i.e. set a 20% upscale limit.

Reply:
It's not very predictable. Even a 5% upscale can visibly degrade the image depending on the upscale algorithm, because no matter how big or small the scale all pixels are affected.

Reply:
Okay then this looks good to me.

Reply:
I would kind of prefer not documenting the vagaries of how we determine this in this way - maybe just as a `//` comment instead of a doc comment.

The concern being that I really think this method is not going to stand up to the test of time, and it'll be "more" breaking since we documented all these implementation details.

Reply:
This is more observable by the developer and their users than a typical implementation detail. Developers are the ones supplying the asset variants and without knowing how we choose them won't know which ones to provide (e.g. https://github.com/flutter/flutter/issues/44108#issuecomment-583292205). I'd prefer not hiding these docs.

> this method is not going to stand up to the test of time

That's fine, we'll just update the docs. I expect we will expand the functionality (e.g. improve physical pixel density detection, support scaled images), but I don't expect much will change otherwise as far as the developer is concerned. Provided there is no extra image scaling, this method should be pretty close to ideal (just like, provided that speeds are small enough and objects big enough, Newtonian physics is close to ideal).